[libc++] Deprecate and remove `uncaught_exception`

frederick-vs-ja opened the pull request:
Works towards P0619R4/#99985.

- `std::uncaught_exception` was not previously deprecated. This patch deprecates it since C++17 as per N4259. `std::uncaught_exceptions` is used instead as libc++ unconditionally provides this function.

Escape hatches:
- `_LIBCPP_ENABLE_CXX20_REMOVED_UNCAUGHT_EXCEPTION` restores `std::uncaught_exception`.

Drive-by changes:
- In Cxx20.rst, says that D.11 is already done. It is done in 578d09c1b195d859ca7e62840ff6bb83421a77b5.

frederick-vs-ja committed:
[libc++] Remove and deprecate `uncaught_exception`

Works towards P0619R4.

- `std::uncaught_exception` was not previously deprecated. This patch deprecates it since C++17 as per N4259. `std::uncaught_exceptions` is used instead as libc++ unconditionally provides this function.

Escape hatches:
- `_LIBCPP_ENABLE_CXX20_REMOVED_UNCAUGHT_EXCEPTION` restores `std::uncaught_exception`.

Drive-by changes:
- In Cxx20.rst, says that D.11 is already done. It is done in 578d09c1b195d859ca7e62840ff6bb83421a77b5.

mordante commented on libcxx/docs/ReleaseNotes/20.rst:
```diff
@@ -44,6 +44,9 @@ Implemented Papers
 Improvements and New Features
 -----------------------------
 
+- The ``_LIBCPP_ENABLE_CXX20_REMOVED_UNCAUGHT_EXCEPTION`` macro has been added to make ``std::uncaught_exception``
+  available.
+
 - TODO
```
```suggestion
- The ``_LIBCPP_ENABLE_CXX20_REMOVED_UNCAUGHT_EXCEPTION`` macro has been added to make ``std::uncaught_exception``
  available.

```

frederick-vs-ja replied:
Done.

mordante commented on libcxx/test/std/language.support/support.exception/uncaught/uncaught_exception.pass.cpp:
```diff
@@ -7,6 +7,10 @@
 
 // UNSUPPORTED: no-exceptions
+
+// ADDITIONAL_COMPILE_FLAGS: -D_LIBCPP_ENABLE_CXX20_REMOVED_UNCAUGHT_EXCEPTION
```
Can you add a new `.verify` test to validate the deprecation?
look at `test/std/utilities/function.objects/negators/unary_negate.depr_in_cxx17.verify.cpp` for an example.

frederick-vs-ja replied:
Added `test/std/language.support/support.exception/uncaught/uncaught_exeption.depr_in_cxx17.verify.cpp`.

mordante reviewed (commented):
Thanks! This looks mostly good, I like to have a quick review after addressing the review comments.

frederick-vs-ja committed:
Address @mordante's review comments

- Remove meaningless `TODO`.
- Add `uncaught_exeption.depr_in_cxx17.verify.cpp`

frederick-vs-ja committed:
Clang-format

frederick-vs-ja committed:
Merge branch 'main' into remove-uncaught_exception

mordante commented on libcxx/modules/std/exception.inc:
```diff
@@ -21,6 +21,5 @@ export namespace std {
   using std::terminate;
   using std::terminate_handler;
   using std::throw_with_nested;
-  using std::uncaught_exception;
```
This should be
```
#ifdef _LIBCPP_ENABLE_CXX20_REMOVED_UNCAUGHT_EXCEPTION
  using std::uncaught_exception;
#endif
```
instead of an unconditional remove.

frederick-vs-ja replied:
Done. But currently we haven't made any other C++17/20/23-removed components restorable in standard library modules. Should we do this later?

mordante reviewed (approved):
Thanks! LGTM modulo one minor issue I missed in the previous review.

frederick-vs-ja committed:
Make `uncaught_exception` restorable in std modules

frederick-vs-ja committed:
Merge branch 'main' into remove-uncaught_exception

ldionne commented on libcxx/docs/Status/Cxx17.rst:
```diff
@@ -40,6 +40,7 @@ Paper Status
 
 .. note::
 
+   .. [#note-N4259] N4259: ``std::uncaught_exception`` is deprecated since version 20.0.
```
```suggestion
   .. [#note-N4259] N4259: ``std::uncaught_exception`` is deprecated since LLVM 20.0.
```

ldionne reviewed (approved):
LGTM with nits.

frederick-vs-ja committed:
Address @ldionne's review comments

Nits in release notes.

frederick-vs-ja committed:
Merge branch 'main' into remove-uncaught_exception

ldionne committed:
Merge branch 'main' into remove-uncaught_exception
